### PR TITLE
 add mock contract to simulate DGX Dex.

### PR DESCRIPTION
**Pull request by talbaneth**

### CODE REVIEW

File: contracts/mockContracts/MockDGXDex.sol [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import "./ecverify.sol";
+
+
+/// @title Mock Digix DEX
+/// @author Tal Baneth
+/// @dev a dummy contract that simulates Digix contract for purchase/sell with verified signatures. 
+contract MockDGXDEX is Withdrawable {
+    uint  constant internal G_TO_MG = (10**3);
+    uint  constant internal DGX_DECIMALS = (10**9);
+
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: we don't use this notation.
only camel back notations...
Dabashot :)s

File: contracts/mockContracts/MockDGXDex.sol [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import "../ERC20Interface.sol";
+import "../Withdrawable.sol";
+import "./ecverify.sol";
+
+
+/// @title Mock Digix DEX
+/// @author Tal Baneth
+/// @dev a dummy contract that simulates Digix contract for purchase/sell with verified signatures. 
+contract MockDGXDEX is Withdrawable {
+    uint  constant internal G_TO_MG = (10**3);
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: can u add some documentation what this means?
or maybe can be clearer names?

File: contracts/mockContracts/MockDGXDex.sol [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            verify_signed_price(block_number,
+                                nonce,
+                                weiPerDgxMg,
+                                signer,
+                                signature);
+
+            require(verified);
+        }
+        if(blockCheck) {
+            require((block_number + 5) >= block.number);
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: this value - 5 - better be configurable...
anyway would rather not use numbers but  const values.d

File: contracts/mockContracts/MockDGXDex.sol [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        require((block_number + 5) >= block.number);
+
+        tokenAmount = (msg.value / weiPerDgxMg) * ( DGX_DECIMALS / G_TO_MG);
+        require(token.transfer(msg.sender, tokenAmount));
+
+        success = true;
+        purchasedAmount = tokenAmount;
+        
+        Purchase(success, purchasedAmount);
+
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: seems like success always true here.
would add to this event more data: buyer address, price
or are u copying digix contract?

File: contracts/mockContracts/MockDGXDex.sol [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        success = true;
+        purchasedAmount = tokenAmount;
+        
+        Purchase(success, purchasedAmount);
+        
+    }
+
+    function sell(uint256 amount,
+                  uint256 block_number,
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: notation.

File: contracts/mockContracts/MockDGXDex.sol [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                  uint256 weiPerDgxMg,
+                  address signer,
+                  bytes signature) public returns (bool success) {
+
+        uint256 amountWei;
+
+        if (sigCheck) {
+            bool verified;
+            address actual_signer;
+
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: notation

File: contracts/mockContracts/MockDGXDex.sol [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                                signature);
+
+            require(verified);
+        }
+        if (blockCheck) {
+            require((block_number + 5) >= block.number);
+        }
+
+        amountWei = amount * G_TO_MG * weiPerDgxMg / DGX_DECIMALS;
+        require(token.transferFrom(msg.sender, this, amount));
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: can't verify this formula. not sure what each number means.

File: contracts/mockContracts/MockDGXDex.sol [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        
+        success = true;
+
+        Sell(success, amountWei);
+    }
+
+    function verify_signed_price(uint _block_number, uint _nonce, uint _price, address _signer, bytes _signature)
+        public
+        returns (bool _verified, address _actual_signer)
+        {
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: { should be indented left.
Also all next lines of this function.

File: contracts/mockContracts/MockDGXDex.sol [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        msg.sender.transfer(amountWei);
+        
+        success = true;
+
+        Sell(success, amountWei);
+    }
+
+    function verify_signed_price(uint _block_number, uint _nonce, uint _price, address _signer, bytes _signature)
+        public
+        returns (bool _verified, address _actual_signer)
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: why do u start here with underscore?

File: contracts/mockContracts/MockDGXDex.sol [hunk 10]
```diff
@@ -0,0 +1,187 @@
+pragma solidity ^0.4.18;
+
+//ask:
+//dst address change?
+//should init another token outside in test?
+
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: yes.
we have testToken contract for creating a token.
u can see it is done in many tests...
just init it and name it digix.

File: contracts/mockContracts/MockDGXDex.sol [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+// blocknumber + 5 >= current block number ? revert
+// signatures - skip for now.
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: require will revert. this if fine.

File: test/mockDGXDEX.js [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        
+        // create dex with the given token. 
+        dgxDex = await mockDGXDEX.new(dgxToken.address, 0, accounts[0]);
+
+        // transfer 100 tokens to dex.
+        let tokenWei = (new BigNumber(10).pow(11)); // 10^9 is one
+        await dgxToken.transfer(dgxDex.address, tokenWei);
+ 
+        // store 10 eth in the dex.
+        let etherWei = new BigNumber(10).pow(19); // 10^18 is one
```
Comment (ilanDoron) [inline test/mockDGXDEX.js]: pow should be 18

File: contracts/mockContracts/MockDGXDex.sol [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    function setBlockCheck (bool _blockCheck) public onlyAdmin {
+        blockCheck = _blockCheck;
+    }
+
+    function purchase(uint256 block_number,
+                      uint256 nonce,
+                      uint256 weiPerDgxMg,
+                      address signer,
+                      bytes signature) payable public returns (bool success, uint256 purchasedAmount) {
+
```
Comment (ilanDoron) [inline contracts/mockContracts/MockDGXDex.sol]: here the { should move to next line.
also payable public ...
if { is not in function line it should move to next line and indented with "function" column

File: contracts/mockContracts/MockDGXDex.sol (hunk 1 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: I do see you use this notation for constants.

File: contracts/mockContracts/MockDGXDex.sol (hunk 2 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: changed to GRAM_TO_MILLIGRAM.

File: contracts/mockContracts/MockDGXDex.sol (hunk 3 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: Done. Added a const.

File: contracts/mockContracts/MockDGXDex.sol (hunk 4 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: The success return (always true) is copied from their contract.
I added more info to the event.

File: contracts/mockContracts/MockDGXDex.sol (hunk 5 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: doen. (the var names were copied from digix, but I changed them to camelcase notation now).

File: contracts/mockContracts/MockDGXDex.sol (hunk 6 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: done.

File: test/mockDGXDEX.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
+var TestToken = artifacts.require("./mockContracts/TestToken.sol");
+var mockDGXDEX = artifacts.require("./mockContracts/MockDGXDEX.sol")
+var Helper = require("./helper.js")
+
+var myToken;
```
Comment (ilanDoron) [inline test/mockDGXDEX.js]: myToken isn't used...

File: test/mockDGXDEX.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+var ethAddress = '0x00eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
+const BigNumber = require('bignumber.js');
+
+contract('MockDGXDEX', function (accounts) {
+    it("create DGX dex and store funds in it.", async function (){
+
+        // create dgx token.
+        dgxToken = await TestToken.new("dex dgx token", "dgx", 9);
+
```
Comment (ilanDoron) [inline test/mockDGXDEX.js]: I think the correct way is to declare this in global scope. and all other globals.
using let.
I see its used during test...

File: contracts/mockContracts/MockDGXDex.sol (hunk 7 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: Changed the calculation to be in steps (also on purchase function). Hope it's clearer now.

Purchase:
amountMg = (msg.value / weiPerDgxMg)
amountMgDgxWei = amountMg * DGX_DECIMALS;
amountDgxWei = amountMgDgxWei / GRAM

Sell:
amountDgxWei = amount;
amountMgDgxWei = amountDgxWei * GRAM_TO_MILLIGRAM;
amountWei = amountMgDgxWei * weiPerDgxMg / DGX_DECIMALS;

File: contracts/mockContracts/MockDGXDex.sol (hunk 8 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: done.

File: contracts/mockContracts/MockDGXDex.sol (hunk 9 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: To distinguish it from storage variables that have the same nam.
I Saw it done in the rest of your contracts.

File: contracts/mockContracts/MockDGXDex.sol (hunk 10 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: done. removed the comments.

File: contracts/mockContracts/MockDGXDex.sol (hunk 11 above)
Comment (talbaneth) [reply contracts/mockContracts/MockDGXDex.sol]: done.

**Comment by talbaneth** [comment]
handled the issues you mentioned.
I did leave the constant notations and the _<var> notation when passing argument with same name as a storage variable. This is same as other contracts.

**Pull request merged**